code/modules/tgui/modal.dm

```diff
@@ -0,0 +1,370 @@
+/**
+ * tgui modals
+ *
+ * Allows creation of modals within tgui.
+ */
+
+GLOBAL_LIST(tgui_modals)
+
+/**
+  * Call this from a proc that is called in tgui_act() to process modal actions
+  *
+  * Example: /obj/machinery/chem_master/proc/tgui_act_modal
+  * You can then switch based on the return value and show different
+  * modals depending on the answer.
+  * Arguments:
+  * * source - The source datum
+  * * action - The called action
+  * * params - The params to the action
+  */
+/proc/tgui_modal_act(datum/source = src, action = "", params)
```

Comment:
Make all these procs `/datum/proc/x` to keep consistency with the other `tgui_` procs